### DISCUSSION TITLE
How is getServerSideProps going to work with React 18 concurrency, streaming and suspense?

### DISCUSSION THREAD

**Discussion by laurencefass** [vercel/next.js#33103] [Help]
I was recently dissapointed to discover that getServerSideProps is synchronous and blocks rendering of components before it completes its internal fetches. Although documented as such there appears to be no way to return a promise to a component in place of data which may take time to retrieve on the server. It doesnt appear to offer any granularity on the partial rendering that Suspense offers.

Do the new React 18 concurrency features make getServerSideProps obsolete? I cant see a use for a blocking initialisation once use of suspense and dynamic imports becomes the norm.

Or does the new Suspense component offer an opportunity to reinvent getServerSideProps to work with Suspense and stream data to its component?

Thanks

**Comment by zanettin** [comment]
Thanks for raising this topic 🙏  remix is planning smth in this direction as well and sounds very promising. you can watch it here https://www.youtube.com/watch?v=BT8cPIAUnck&t=3156s.
start streaming immediately and let the client download assets via Link headers right away and then start to execute the `getServerSideProps` part and stream that on completion as well. so no waiting for the resolving of the promise. the only issue on this solution could be, that when streaming has started and you'd like to modify/add headers on the `getServerSideProps` handler based on an API response or so, you wouldn't be able to do so. imo the streaming should be configurable by the dev to tell the framework if we'd like to make use of the streaming or not. just my thoughts about it.

**Reply by laurencefass** [reply]
thanks.

**Comment by laurencefass** [comment]
Does anyone have any insights into how this would work with Next JS?

Its a major useability issue having to wait for the server to load and offsets the benefits of being able to do it if the entire page is delayed waiting for data on the backend. On paper at least Suspense makes it also unnecessary.

Would it be possible to make getServerSideProps compatible with Suspense so to allow use of fallbacks in _app.js?

Related to this: why is gssp an async function if returning a promise has no effect? Is there  a use-case for returning a promise I have missed?

Could Next JS convert that promise into a async callback or hook across the network as its implementation of the Suspense API?
